### PR TITLE
Support parallel formatting almost for all text formats

### PR DESCRIPTION
**Pull request by Avogar**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support parallel formatting for all text formats, except  `JSONEachRowWithProgress` and `PrettyCompactMonoBlock`

Detailed description / Documentation draft:
...

### CODE REVIEW

**Comment by Avogar** [comment]
@Mergifyio update

File: src/Processors/Formats/IOutputFormat.h [hunk 1]
```diff
@@ -85,6 +95,55 @@ class IOutputFormat : public IProcessor
...
+        if (need_write_suffix)
+        {
+            writeSuffix();
+            need_write_suffix = false;
+        }
+    }
+
+    /// Methods-helpers for parallel formatting.
+
+    /// Set the real number of first row in the first chunk of data.
```
Comment (nikitamikhaylov) [inline src/Processors/Formats/IOutputFormat.h:109]: But it is not really for the first chunk of data, but for every chunk of data?

File: src/Processors/Formats/IOutputFormat.h [hunk 2]
```diff
@@ -85,6 +95,55 @@ class IOutputFormat : public IProcessor
...
+
+    /// In some formats the way we print extremes depends on
+    /// were totals printed or not. In this case in parallel formatting
+    /// we should notify underling format if totals were printed.
+    void setTotalsAreWritten() { are_totals_written = true; }
+    bool areTotalsWritten() const { return are_totals_written; }
+
+    /// Return true if format saves totals and extremes in consumeTotals/consumeExtremes and
+    /// outputs them in finalize() method.
+    virtual bool areTotalsAndExtremesUsedInFinalize() const { return false; }
```
Comment (nikitamikhaylov) [inline src/Processors/Formats/IOutputFormat.h:145]: Why do we need this? Aren't your code generic enough? Underlying format may decide for itself what to do with the statistics?

Comment (nikitamikhaylov) [reply src/Processors/Formats/IOutputFormat.h:145]: UPD: I understood

**Review by nikitamikhaylov — APPROVED** [review]
The code became really complicated. It took me much time to understand all caveats and pitfalls. But i hope I it is Ok tradeoff in return for performance.

File: src/Processors/Formats/IOutputFormat.h (hunk 1 above)
Comment (Avogar) [reply src/Processors/Formats/IOutputFormat.h:109]: Not exacetely, the first row number in the second chunk is just the first row number in the first chunk + the number of rows in the first chunk, and etc. Maybe just bad comment or field/method name (will think about better naming)

File: src/Processors/Formats/IOutputFormat.h (hunk 2 above)
Comment (Avogar) [reply src/Processors/Formats/IOutputFormat.h:145]: The main problem is that in parallel formatting we use separate threads for consumeTotals/consumeExtremes/finalize (and it's not bad), but some formats can save some state in these methods and use it in the other method. I haven't come up with anything better than just save such states in ParallelFormattingOutputFormat and set it to formatters in threads. It made code a bit tricky and complicated, but I thought it worth it (and maybe we will try to refactor it in future).

File: src/Processors/Formats/IOutputFormat.h (hunk 1 above)
Comment (nikitamikhaylov) [reply src/Processors/Formats/IOutputFormat.h]: I see

**Pull request merged**